### PR TITLE
Supporting llama uint4 quantization using AutoGPTQ

### PR DESCRIPTION
**Pull request by HolyFalafel**

### CODE REVIEW

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+
+import numpy as np
+import torch
+import torch.nn as nn
+import transformers
+import habana_frameworks.torch.core as htcore
+
+
+logger = getLogger(__name__)
+# try:
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:12]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+        infeatures,
+        outfeatures,
+        bias,
+        use_cuda_fp16=True,
+        kernel_switch_threshold=128,
+        trainable=False,
+        weight_dtype=torch.float16,
+    ):
+        print(f"in qlinear_hpu : QuantLinear::__init__")
+        print(f"__init__ {bits=}, {group_size=}, {infeatures=}, {outfeatures=}, {bias=}, {use_cuda_fp16=}, {kernel_switch_threshold=}, {trainable=}, {weight_dtype=}")
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:40]: use the logger

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+        super().__init__()
+        # global _autogptq_cuda_available
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:42]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        if bits not in [2, 3, 4, 8]:
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:43]: we only support 4 bit

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+            raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+        # if trainable:
+        #     _autogptq_cuda_available = False
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:46]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+            torch.tensor([i // self.group_size for i in range(infeatures)], dtype=torch.int32),
+        )
+
+        if bias:
+            self.register_buffer("bias", torch.zeros((outfeatures), dtype=weight_dtype))
+        else:
+            self.bias = None
+        self.half_indim = self.infeatures // 2
+
+        self.use_hpu_fp16 = use_cuda_fp16 if bits != 8 else False
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:85]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+
+        # is performed by unpacking the weights and using torch.matmul
+        if self.bits in [2, 4, 8]:
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:88]: no need to check as we support only 4

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+            self.wf = torch.tensor(list(range(0, 32, self.bits)), dtype=torch.int32).unsqueeze(0)
+        elif self.bits == 3:
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:90]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+                [
+                    [0, 3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 0],
+                    [0, 1, 4, 7, 10, 13, 16, 19, 22, 25, 28, 31],
+                    [0, 2, 5, 8, 11, 14, 17, 20, 23, 26, 29, 0],
+                ],
+                dtype=torch.int32,
+            ).reshape(1, 3, 12)
+
+        self.kernel_switch_threshold = kernel_switch_threshold
+        # self.autogptq_cuda_available = _autogptq_cuda_available
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:101]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+                col += 1
+            else:
+                raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+
+        qzeros = qzeros.astype(np.int32)
+        self.qzeros = torch.from_numpy(qzeros)
+
+    def forward(self, x):
+        print(f"forward {x.shape=}")
+        print(f"forward {self.scales=} {self.qzeros=}")
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:207]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+        x_dtype = x.dtype
+        out_shape = x.shape[:-1] + (self.outfeatures,)
+        x = x.reshape(-1, x.shape[-1])
+        # if (
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:211]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+        #     x.device.type == "cuda"
+        #     and self.autogptq_cuda_available is True
+        #     and (self.kernel_switch_threshold is False or x.shape[0] < self.kernel_switch_threshold)
+        # ):
+        if False:
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:216]: remove

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+                        self.qzeros,
+                        self.group_size,
+                    )
+                else:
+                    raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+        else:
+            if self.wf.device != self.qzeros.device:
+                self.wf = self.wf.to(self.qzeros.device)
+
+            if self.bits in [4]:
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:301]: remove this if

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+                    # qzeros = qzeros.reshape(qzeros.shape[-1], -1)
+                    # scales = scales.reshape(scales.shape[-1], -1)
+                    qzeros = qzeros.reshape(-1, qzeros.shape[-1]).contiguous()
+                    scales = scales.reshape(-1, scales.shape[-1]).contiguous()
+                    print(f"forward {scales.shape=} {qzeros.shape=}")
+                weight = torch.ops.hpu.convert_from_uint4(qweight, scales, qzeros, x_dtype)
+                print(f"QuantLinear::forward {x.shape=} {weight.shape=} {weight.shape[-1]//8=}")
+                weight = weight.reshape(-1, weight.shape[-1]//8)
+                print(f"QuantLinear::forward {weight.shape=}")
+            else:
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:331]: remove

File: auto_gptq/utils/import_utils.py [hunk 15]
```diff
@@ -66,6 +66,14 @@ def dynamically_import_QuantLinear(
     use_qigen: bool = False,
     disable_marlin: bool = True,
 ):
+    # if use_hpu:
+    try:
+        import habana_frameworks.torch.hpu  # noqa: F401
+    except ImportError as e:
+        pass
+    else:
```
Comment (MrGeva) [inline auto_gptq/utils/import_utils.py:74]: else of what?

**Review by MrGeva — COMMENTED** [review]
please cleanup the code, its hard to review like that

File: auto_gptq/utils/import_utils.py (hunk 15 above)
Comment (HolyFalafel) [reply auto_gptq/utils/import_utils.py:74]: Of the try-except

The else block lets you execute code when there is no error.

https://www.w3schools.com/python/python_try_except.asp

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+        self.qweight = new_qweight.to('hpu')
+
+        zeros = self.unpack_zeros_cuda().cpu()
+        new_qzeros = pack_cuda_old_hpu(zeros)
+        self.qzeros = new_qzeros.to('hpu')
+
+    def post_init(self):
+        self._preprocessing()
+
+    def pack(self, linear, scales, zeros, g_idx):
```
Comment (Yantom1) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:98]: Im not sure what is this pack used for if do everything in preprocess

File: tests/test_q4.py [hunk 17]
```diff
@@ -6,12 +6,13 @@
...
 
 
-try:
-    from exllama_kernels import prepare_buffers, set_tuning_params
-except ImportError as e:
-    print(f"[WARNING] Could not load exllama_kernels: {e}")
+# try:
+#     from exllama_kernels import prepare_buffers, set_tuning_params
+# except ImportError as e:
+#     print(f"[WARNING] Could not load exllama_kernels: {e}")
```
Comment (Yantom1) [inline tests/test_q4.py:15]: remove?

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py (hunk 16 above)
Comment (HolyFalafel) [reply auto_gptq/nn_modules/qlinear/qlinear_hpu.py:98]: This is for the quantization process, we're still not there.
Currently it's copied from cuda_old.
We'll need to change it if we quantize in AutoGPTQ

Comment (MrGeva) [reply auto_gptq/nn_modules/qlinear/qlinear_hpu.py:98]: let's remove this code if we are not sure it is working. and throw unsupported exception.

File: auto_gptq/nn_modules/qlinear/qlinear_hpu.py [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+        out_shape = x.shape[:-1] + (self.outfeatures,)
+        x = x.reshape(-1, x.shape[-1])
+        scales = self.scales
+        qweight = self.qweight
+        zeros = self.qzeros
+        weight = torch.ops.hpu.convert_from_uint4(qweight, scales, zeros, x_dtype)
+        out = torch.matmul(x, weight)
+        out = out.to(dtype=x_dtype).reshape(
+            out_shape
+        )  # A cast is needed here as for some reason the vecquant2matmul_faster_old still allocate a float32 output.
```
Comment (MrGeva) [inline auto_gptq/nn_modules/qlinear/qlinear_hpu.py:201]: is the .to() relevant for us?

**Review by ulivne — CHANGES_REQUESTED** [review]
Suggest to rename the PR to uint4 instead int4, as int4 is not fully verifred (no refernece),
right?

**Review by HolyFalafel — COMMENTED** [review]
Now see replies

**Review by ulivne — APPROVED** [review]

**Pull request merged**